### PR TITLE
Upgrade gosnmp to support more authentication and privacy protocols for v3 connections

### PR DESCRIPTION
**Pull request by ofek**
### Motivation

- https://github.com/soniah/gosnmp/blob/b45a3945991960546f95021d9144ea611c321494/v3_usm.go#L31-L39
- https://github.com/soniah/gosnmp/blob/b45a3945991960546f95021d9144ea611c321494/v3_usm.go#L112-L120

### CODE REVIEW

**Review by truthbk — APPROVED** [review]
Nice!

File: pkg/snmp/snmp.go [hunk 1]
```diff
@@ -144,20 +144,40 @@ func (c *Config) BuildSNMPParams() (*gosnmp.GoSNMP, error) {
 
 	var authProtocol gosnmp.SnmpV3AuthProtocol
 	lowerAuthProtocol := strings.ToLower(c.AuthProtocol)
-	if lowerAuthProtocol == "md5" {
+	if lowerAuthProtocol == "none" {
```
Comment (AlexandreYang) [inline pkg/snmp/snmp.go:147]: (minor) Is `lowerAuthProtocol == "none"` needed ? Doesn't **not setting** the option already mean NoAuth?

Same comment for the other case.

File: pkg/snmp/snmp.go [hunk 2]
```diff
@@ -144,20 +144,40 @@ func (c *Config) BuildSNMPParams() (*gosnmp.GoSNMP, error) {
...
+		authProtocol = gosnmp.SHA224
+	} else if lowerAuthProtocol == "sha256" {
+		authProtocol = gosnmp.SHA256
+	} else if lowerAuthProtocol == "sha384" {
+		authProtocol = gosnmp.SHA384
+	} else if lowerAuthProtocol == "sha512" {
+		authProtocol = gosnmp.SHA512
 	} else if c.AuthProtocol != "" {
 		return nil, fmt.Errorf("Unsupported authentication protocol: %s", c.AuthProtocol)
 	}
```
Comment (FlorianVeaux) [inline pkg/snmp/snmp.go:163]: Continuing on that comment, `gosnmp.NoAuth` equals 1: https://godoc.org/github.com/soniah/gosnmp#SnmpV3AuthProtocol

Further down we set `msgFlags = gosnmp.AuthNoPriv` https://github.com/DataDog/datadog-agent/pull/5632/files#diff-e3d79654e541e8cdc1941ea637ff3de3R188 on an non-empty lowerAuthProtocol
gosnmp will raise an error if a user sets `authentication_protocol: none` in their config https://github.com/soniah/gosnmp/blob/master/v3_usm.go#L267

I think the correct approach would be something like this:
```suggestion
	if lowerAuthProtocol == "" {
		authProtocol = gosnmp.NoAuth
	} else if lowerAuthProtocol == "md5" {
		authProtocol = gosnmp.MD5
	} else if lowerAuthProtocol == "sha" {
		authProtocol = gosnmp.SHA
	} else if lowerAuthProtocol == "sha224" {
		authProtocol = gosnmp.SHA224
	} else if lowerAuthProtocol == "sha256" {
		authProtocol = gosnmp.SHA256
	} else if lowerAuthProtocol == "sha384" {
		authProtocol = gosnmp.SHA384
	} else if lowerAuthProtocol == "sha512" {
		authProtocol = gosnmp.SHA512
	} else {
		return nil, fmt.Errorf("Unsupported authentication protocol: %s", c.AuthProtocol)
	}
```

Same comment for privacy protocol

**Review by FlorianVeaux — CHANGES_REQUESTED** [review]
See: https://github.com/DataDog/datadog-agent/pull/5632/files#r431733911

**Review by FlorianVeaux — APPROVED** [review]
💯

**Pull request merged**